File: LeetCode/Medium/0814-binary-tree-pruning/0814-binary-tree-pruning.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* pruneTree(TreeNode* root) {
        DFS(root);
        
        if (root->val == 0 && !root->left && !root->right) return NULL;
        return root;
    }
    
    bool DFS(TreeNode* node){
        if (!node) return false;
        
        bool left = DFS(node->left);
        bool right = DFS(node->right);
        
        if (!left) node->left = NULL;
        if (!right) node->right = NULL;
        
        return node->val == 1 || left || right;
    }
};

File: LeetCode/Medium/0814-binary-tree-pruning/0814-binary-tree-pruning.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode pruneTree(TreeNode root) {
        DFS(root);
        
        if (root.val == 0 && root.left == null && root.right == null) return null;
        return root;
    }
    
    private boolean DFS(TreeNode node){
        if (node == null) return false;
        
        boolean left = DFS(node.left);
        boolean right = DFS(node.right);
        
        if (!left) node.left = null;
        if (!right) node.right = null;
        
        return node.val == 1 || left || right;
    }
}